EMA: An aligner for barcoded short-read sequencing data
=======================================================
[![Build Status](https://travis-ci.org/arshajii/ema.svg?branch=master)](https://travis-ci.org/arshajii/ema) [![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/arshajii/ema/master/LICENSE)

EMA uses a latent variable model to align barcoded short-reads (such as those produced by [10x Genomics](https://www.10xgenomics.com)' sequencing platform). More information is available in [our paper](https://www.biorxiv.org/content/early/2017/11/16/220236). The full experimental setup is available [here](https://github.com/arshajii/ema-paper-data/blob/master/experiments.ipynb).

### Download and Compile
In a nutshell:

```
git clone --recursive https://github.com/arshajii/ema
cd ema
make
```

The `--recursive` flag is needed because EMA uses BWA's C API.

### Usage
```
usage: ./ema <preproc|count|align|help> [options]

preproc: preprocess barcoded FASTQ files
  -w <whitelist path>: specify whitelist [required]
  -n <num buckets>: number of barcode buckets to make [500]
  -h: apply Hamming-2 correction [off]
  -o: <output directory> specify output directory [required]
  -t <threads>: set number of threads [1]
  All other arguments: list of *all* output prefixes generated by count stage

count: performs preliminary barcode count (takes FASTQ via stdin)
  -w <whitelist path>: specify barcode whitelist [required]
  -o <output prefix>: specify output prefix [required]

align: choose best alignments based on barcodes
  -1 <FASTQ1 path>: first (preprocessed and sorted) FASTQ file [required]
  -2 <FASTQ2 path>: second (preprocessed and sorted) FASTQ file [required]
  -s <EMA-FASTQ path>: specify special FASTQ path [none]
  -x: multi-input mode; takes input files after flags and spawns a thread for each
  -r <FASTA path>: indexed reference [required]
  -o <SAM file>: output SAM file [stdout]
  -R <RG string>: full read group string (e.g. '@RG\tID:foo\tSM:bar') [none]
  -d: apply fragment read density optimization
  -p <platform>: sequencing platform (one of '10x', 'tru', 'cpt') [10x]
  -t <threads>: set number of threads [1]

help: print this help message
```

### Preprocessing (10x only)
_TODO_
Instructions for preprocessing and running EMA on data from other sequencing platforms can be found [here](https://github.com/arshajii/ema-paper-data/blob/master/experiments.ipynb).

### Input formats
EMA has several input modes:
- `-s <input>`: Input file is a single preprocessed "special" FASTQ generated by the preprocessing steps above.
- `-x`: Input files are listed after flags (as in `ema align -a -b -c <input 1> <input 2> ... <input N>`). Each of these inputs are processed and all results are written to the SAM file specified with `-o`.
- `-1 <first mate>`/`-2 <second mate>`: Input files are standard FASTQs. For interleaved FASTQs, `-2` can be omitted. The only restriction in this input mode is that read identifiers must end in `:<barcode sequence>`. 

### Parallelism
Multithreading can be enabled with `-t <num threads>`. The actual threading mode is dependent on how the input is being read, however:
- `-s`, `-1`/`-2`: Multiple threads are spawned to work on the single input file (or pair of input files).
- `-x`: Threads work on the input files individually.

(Note that, because of this, it never makes sense to spawn more threads than there are input files when using `-x`.)

### EMA wrapper
A wrapper script is provided in [util/ema_wrapper.sh](util/ema_wrapper.sh), and can be invoked like this:

```
EMAPATH=/path/to/ema/executable ./ema_wrapper.sh
```

### Output
EMA outputs a standard SAM file with several additional tags:

- `XG`: alignment probability
- `XC`: cloud identifier
- `XA`: alternate high-probability alignments
